Solutions/Workshops/workshop6.lean

```diff
@@ -0,0 +1,284 @@
+
+-- Use of the `@[ext]` attribute means Lean generates theorems for proving things by extensionality
+#check Point.ext
+
+example (a b : Point ℕ ) (hx : a.x = b.x) (hy : a.y = b.y) : a = b := by
+  ext
+  repeat' assumption
+
+-- We can print all the details of a structure using `#print prefix`.
+#print prefix Point
+
+-- Notice that the inductive type version has fewer supporting functions.
+#print prefix Point'
+
+-- There are various ways to define instances of a structure.
+-- By providing the named fields directly:
+#check { x := 10, y := 20 : Point ℕ }
+#check { y := 20, x := 10 : Point ℕ }
+#check ({x := 10, y := 20} : Point ℕ)
+
+-- Using the anonymous constructor:
+#check (⟨10, 20⟩ : Point ℕ)
+
+-- Using the default constructor:
+#check (Point.mk 10 20)
+
+-- The anonymous constructor and named fields are useful for defining functions on structures.
+
+namespace Point -- any structure creates a namespace
+
+def add : Point ℕ → Point ℕ → Point ℕ :=
+  λ a b => ⟨a.x + b.x, a.y + b.y⟩
+
+def smul (n : ℕ) (p : Point ℕ) :=
+  Point.mk (n * p.x) (n * p.y)
+
+end Point
+
+#check @Point.add
+#check @Point.smul
+
+def p := Point.mk 1 2
+def q := Point.mk 3 4
+
+-- Notice the order in which arguments are input to functions here.
+#eval p.add q
+#check p.add q
+
+#eval p.smul 3
+#check p.smul 3
+
+-- We can extend structures by adding new fields. Here we extend a `Point` by adding RGB values.
+
+structure RGBValue where
+  red : ℕ
+  green : ℕ
+  blue : ℕ
+
+structure RGBPoint (α : Type*) extends Point α, RGBValue
+
+#print prefix RGBPoint
+#check RGBPoint.mk
+#check RGBPoint.toPoint
+#check RGBPoint.toRGBValue
+def cp := RGBPoint.mk (Point.mk 1 2) (RGBValue.mk 10 20 30)
+
+#eval cp.x
+#eval cp.red
+
+-- We can use the fields of a structure to specify constraints that must be satisfied.
+
+structure RedGreenPoint (α : Type*) extends Point α, RGBValue where
+  no_blue : blue = 0
+
+#check RedGreenPoint.mk
+
+def rgp : RedGreenPoint ℕ :=
+  { p with red := 200, green := 40, blue := 0, no_blue := rfl }
+
+example : rgp.x   = 1 := rfl
+example : rgp.red = 200 := rfl
+
+-- Here is an example of the use of fields in a structure to give contraints.
+
+structure Group₁ (α : Type*) where
+  mul : α → α → α
+  one : α
+  inv : α → α
+  mul_assoc : ∀ x y z : α, mul (mul x y) z = mul x (mul y z)
+  mul_one : ∀ x : α, mul x one = x
+  one_mul : ∀ x : α, mul one x = x
+  mul_left_inv : ∀ x : α, mul (inv x) x = one
+
+/-!
+# Exercises
+-/
+
+-- 1. Prove associativity of addition on `Point ℕ`. [MIL 6.1]
+
+namespace Point
+
+theorem add_assoc (a b c : Point ℕ) : (a.add b).add c = a.add (b.add c) := by
+{ simp only [add, Point.ext_iff, Nat.add_assoc]
+  tada }
+
+end Point
+
+-- 2. Here we define a type of special RGB points.
+
+structure SpecialPoint extends Point ℕ, RGBValue where
+  gradientEqn : x = red
+
+-- Now define an addition function for special points.
+
+namespace SpecialPoint
+
+def add : SpecialPoint → SpecialPoint → SpecialPoint := λ a b ↦ sorry
```

Comment:
not sure how to fill this `sorry` in

Reply:
Yeah, that's what I was thinking too. I guess we can just do like what you said and add a note above it?